AG-22 - Add Agent Tests

SammyOina opened the pull request:
### What does this do?
Adds tests to agent 

### Which issue(s) does this PR fix/relate to?
Resolves #22 

### List any changes that modify/break current functionality

### Have you included tests for your changes?

### Did you document any new/modified functionality?

### Notes

SammyOina committed:
```
Add tests for NewHeartbeat and Update functions
```

This commit adds tests for the `NewHeartbeat` and `Update` functions in the `agent` package. The `NewHeartbeat` test checks if the `NewHeartbeat` function correctly initializes a new heartbeat with the given name, service type, and interval. It also verifies the initial status and info of the heartbeat.

Signed-off-by: SammyOina <sammyoina@gmail.com>

SammyOina committed:
extract constants

Signed-off-by: SammyOina <sammyoina@gmail.com>

SammyOina committed:
Add bootstrap tests.

Fix error when writing to TOML file

The commit fixes an error that occurs when writing to a TOML file.
The error is wrapped with a custom error message.
This ensures that the error is properly handled and provides more
context to the developer.

The commit also includes changes to the import statements in the
bootstrap.go file.

Signed-off-by: SammyOina <sammyoina@gmail.com>

SammyOina committed:
Fix error handling in bootstrap_test.go

The commit fixes error handling in the bootstrap_test.go file by checking for an error when writing the response body. This ensures that any errors are properly handled and reported.

Signed-off-by: SammyOina <sammyoina@gmail.com>

SammyOina committed:
Fix context bug in NewHeartbeat function to allow graceful stop to listen function

Signed-off-by: SammyOina <sammyoina@gmail.com>

SammyOina committed:
Fix race condition in Heartbeat.Info() method

The `Info()` method in the `Heartbeat` interface was not thread-safe,
which could lead to a race condition when accessing the `info` field.
To fix this issue, a mutex was added to protect the access to `info`
and ensure thread safety. The `Info()` method now locks the mutex
before accessing `info` and unlocks it before returning the value.
Additionally, a new `Close()` method was added to stop the heartbeat
ticker when it is no longer needed.

Signed-off-by: SammyOina <sammyoina@gmail.com>

SammyOina committed:
Fix error handling in agent/config.go

The commit fixes error handling in agent/config.go
by adding more specific error variables and updating
the error messages. This improves the clarity and
readability of the codebase.

Signed-off-by: SammyOina <sammyoina@gmail.com>

SammyOina committed:
test ci

Signed-off-by: SammyOina <sammyoina@gmail.com>

SammyOina committed:
test push

Signed-off-by: SammyOina <sammyoina@gmail.com>

SammyOina committed:
test 22

Signed-off-by: SammyOina <sammyoina@gmail.com>

SammyOina committed:
test target

Signed-off-by: SammyOina <sammyoina@gmail.com>

SammyOina committed:
switch context

Signed-off-by: SammyOina <sammyoina@gmail.com>

SammyOina committed:
switch context

Signed-off-by: SammyOina <sammyoina@gmail.com>

SammyOina committed:
make agent

Signed-off-by: SammyOina <sammyoina@gmail.com>

SammyOina committed:
debug build

Signed-off-by: SammyOina <sammyoina@gmail.com>

SammyOina committed:
build args

Signed-off-by: SammyOina <sammyoina@gmail.com>

SammyOina committed:
Add terminal tests

Refactor build workflow

The build22.yml file has been deleted. This commit removes the build22.yml file, which was used to create and publish a Docker image. The file is no longer needed as it is no longer being used in the project.

Signed-off-by: SammyOina <sammyoina@gmail.com>

SammyOina committed:
Add test for PushOperation function in edgex package

This commit adds a new test for the PushOperation function in the edgex package. The test creates a mock HTTP server and checks the HTTP request method, URL, and request body. The test ensures that the function correctly handles the POST request with the expected URL and request body.

Signed-off-by: SammyOina <sammyoina@gmail.com>

SammyOina committed:
Add tests for the conn package

This commit adds tests for the conn package. The tests cover various scenarios and ensure the functionality of the package is working as expected.

Signed-off-by: SammyOina <sammyoina@gmail.com>

SammyOina committed:
Fix bug in client_test.go

The commit fixes a bug in the client_test.go file where the expected response was not being properly set. The bug caused the test to fail. The fix ensures that the expected response is correctly set and the test passes as expected.

Signed-off-by: SammyOina <sammyoina@gmail.com>

SammyOina committed:
Fix error handling in main.go and endpoints_test.go

The commit fixes error handling in main.go and endpoints_test.go.
It ensures that the server is properly shutdown and logs any errors
that occur during the shutdown process.

Signed-off-by: SammyOina <sammyoina@gmail.com>

rodneyosodo commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,33 @@
+on:
+  push:
+    branches:
+      - tests  # Change this to your main branch name if different
```
Remove comments

rodneyosodo commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,33 @@
+
+    steps:
+      - name: Checkout Repository
+        uses: actions/checkout@v2
```
```suggestion
        uses: actions/checkout@v4
```

rodneyosodo commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,33 @@
+        uses: actions/checkout@v2
+
+      - name: Set up Go
+        uses: actions/setup-go@v2
```
```suggestion
        uses: actions/setup-go@v4
```

rodneyosodo commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,33 @@
+        run: |
+          make dockers
+
+      - name: Docker Login
```
You can use https://github.com/marketplace/actions/build-and-push-docker-images

rodneyosodo commented on Makefile:
```diff
@@ -37,7 +37,7 @@ define make_docker
 		--build-arg COMMIT=$(COMMIT) \
 		--build-arg TIME=$(TIME) \
-		--tag=mainflux/$(svc) \
+		--tag=sammyoina/$(svc) \
```
Revert

rodneyosodo commented on Makefile:
```diff
@@ -82,7 +82,7 @@ dockers_dev: $(DOCKERS_DEV)
 define docker_push
 	for svc in $(SERVICES); do \
-		docker push mainflux/$$svc:$(1); \
+		docker push sammyoina/$$svc:$(1); \
```
Same

rodneyosodo commented on docker/Dockerfile:
```diff
@@ -10,8 +10,9 @@ WORKDIR /go/src/github.com/mainflux/agent
-    && make $SVC \
-    && mv build/mainflux-$SVC /exe
+    && make $SVC 
+RUN ls build
```
Revert

rodneyosodo commented on docker/Dockerfile:
```diff
@@ -10,8 +10,9 @@ WORKDIR /go/src/github.com/mainflux/agent
-    && mv build/mainflux-$SVC /exe
+    && make $SVC 
+RUN ls build
+RUN mv build/mainflux-$SVC /exe
```
This should be left combined to reduce number of layers

rodneyosodo commented on pkg/bootstrap/bootstrap_test.go:
```diff
@@ -0,0 +1,140 @@
+			assert.True(t, errors.Contains(err, test.expectedErr), fmt.Sprintf("expected %v got %v", test.expectedErr, err))
+		})
+	}
+	// cleanup.
```
Remove comment

rodneyosodo commented on pkg/edgex/client_test.go:
```diff
@@ -0,0 +1,165 @@
+	// Create an edgexClient with the mock server URL
+	client := NewClient(server.URL+"/", logger.NewMock())
+
+	// Test PushOperation
```
Remove unnecessary comment as some part of the code is self documenting

rodneyosodo commented on pkg/edgex/client_test.go:
```diff
@@ -0,0 +1,165 @@
+
+func TestFetchConfig(t *testing.T) {
+
+	// Create a mock HTTP server
```
Same

SammyOina committed:
remove test.yml

Signed-off-by: SammyOina <sammyoina@gmail.com>

SammyOina committed:
Fix Docker image tag and file path in Makefile and Dockerfile

The Docker image tag and file path in the Makefile and Dockerfile have been fixed to use the correct repository name "Mainflux" instead of "sammyoina". This ensures that the Docker images are tagged and pushed to the correct repository. Additionally, the file path in the Dockerfile has been updated to correctly move the built executable to the desired location.

Signed-off-by: SammyOina <sammyoina@gmail.com>

SammyOina committed:
remove comment

Signed-off-by: SammyOina <sammyoina@gmail.com>

SammyOina committed:
Fix Docker image tag in Makefile

The Docker image tag in the Makefile was updated to use lowercase "mainflux" instead of "Mainflux". This ensures consistency with the naming convention used for the Docker images.

Signed-off-by: SammyOina <sammyoina@gmail.com>

rodneyosodo reviewed (dismissed):
LGTM

dborovcanin commented on pkg/bootstrap/bootstrap_test.go:
```diff
@@ -0,0 +1,139 @@
+					}
+				  }
+				],
+				"content": "{\"agent\":{\"edgex\":{\"url\":\"http://localhost:48090/api/v1/\"},\"heartbeat\":{\"interval\":\"30s\"},\"log\":{\"level\":\"debug\"},\"mqtt\":{\"mtls\":false,\"qos\":0,\"retain\":false,\"skip_tls_ver\":true,\"url\":\"tcp://mainflux-domain.com:1883\"},\"server\":{\"nats_url\":\"localhost:4222\",\"port\":\"9000\"},\"terminal\":{\"session_timeout\":\"30s\"}},\"export\":{\"exp\":{\"c ...
```
Extract the content to a separate multiline string so you avoid escaping and make it a little more readable. Use `fmt.Sprintf` to construct a response with content.

dborovcanin commented on pkg/terminal/terminal_test.go:
```diff
@@ -0,0 +1,69 @@
+		t.Fatalf("Expected no error, but got: %v", err)
+	}
+
+	// Simulate writing data to the session
```
We should add tests unhappy flow to test error handling.

dborovcanin replied:
The same for other methods.

SammyOina committed:
add test cases and format string

Signed-off-by: SammyOina <sammyoina@gmail.com>
